### PR TITLE
🐛 Fix amp-list race condition

### PR DESCRIPTION
**Pull request by samouri**
**summary**
Addresses https://github.com/ampproject/amphtml/issues/25961

Adds a check that the `src` has not changed while the xhr was inflight. If the `src` has changed, then bail instead of scheduling a render.  I've added a test that previously would have resulted in two renders and now instead only has one.

cc @jridgewell  / @choumx

### CODE REVIEW

**Comment by samouri** [comment]
@choumx  / @jridgewell: PTAL

File: extensions/amp-list/0.1/test/test-amp-list.js [hunk 1]
```diff
@@ -895,6 +895,32 @@ describes.repeated(
...
+            const foo = doc.createElement('div');
+            const bar = doc.createElement('div');
+
+            const firstPromise = Promise.resolve({items: [foo]});
+            const secondPromise = firstPromise.then(() => ({items: [bar]}));
+
+            listMock
+              .expects('fetch_')
+              .returns(firstPromise)
+              .returns(secondPromise)
```
Comment (choumx) [inline extensions/amp-list/0.1/test/test-amp-list.js]: Does this `returns()` make the line above noop? 

You might need `onCall(n).returns(m)`: https://sinonjs.org/releases/v7.5.0/stubs/

File: extensions/amp-list/0.1/test/test-amp-list.js [hunk 2]
```diff
@@ -895,6 +895,32 @@ describes.repeated(
...
+              .twice();
+            listMock
+              .expects('scheduleRender_')
+              .withArgs([bar])
+              .returns(Promise.resolve())
+              .once();
+            listMock.expects('showFallback_').never();
+
+            element.setAttribute('src', 'https://foo.com/list.json');
+            list.layoutCallback();
```
Comment (choumx) [inline extensions/amp-list/0.1/test/test-amp-list.js]: Maybe add a comment that you're intentionally not awaiting this promise and why.

File: extensions/amp-list/0.1/test/test-amp-list.js [hunk 3]
```diff
@@ -895,6 +895,32 @@ describes.repeated(
             expect(element.getAttribute('src')).to.equal('');
           });
 
+          it('should not render if [src] has changed since the fetch was initiated', async () => {
```
Comment (choumx) [inline extensions/amp-list/0.1/test/test-amp-list.js]: Does this test fail if you revert the changes in amp-list.js?

File: extensions/amp-list/0.1/amp-list.js [hunk 4]
```diff
@@ -545,14 +545,20 @@ export class AmpList extends AMP.BaseElement {
...
       return Promise.resolve();
     }
     let fetch;
     if (this.ssrTemplateHelper_.isSupported()) {
       fetch = this.ssrTemplate_(opt_refresh);
     } else {
       fetch = this.prepareAndSendFetch_(opt_refresh).then(data => {
+        // Bail if the src has changed while resolving the xhr request.
+        if (elementSrc !== this.element.getAttribute('src')) {
+          return Promise.resolve();
```
Comment (jridgewell) [inline extensions/amp-list/0.1/amp-list.js]: Just return, no need to create a promise.

File: extensions/amp-list/0.1/amp-list.js [hunk 5]
```diff
@@ -678,7 +685,13 @@ export class AmpList extends AMP.BaseElement {
           throw user().createError('Error proxying amp-list templates', error);
         }
       )
-      .then(data => this.scheduleRender_(data, /* append */ false));
+      .then(data => {
+        // Bail if the src has changed while resolving the xhr request.
+        if (elementSrc !== this.element.getAttribute('src')) {
+          return Promise.resolve();
```
Comment (jridgewell) [inline extensions/amp-list/0.1/amp-list.js]: Ditto.

File: extensions/amp-list/0.1/test/test-amp-list.js (hunk 1 above)
Comment (samouri) [reply extensions/amp-list/0.1/test/test-amp-list.js]: You are right! Can't just assume that things that work in [truth](https://github.com/google/truth) also work here :). 

I'm guessing this test still worked anyway since the code didn't have any optimizations to bail even though the result was the same (at least not as high up as i'm testing)

File: extensions/amp-list/0.1/test/test-amp-list.js (hunk 3 above)
Comment (samouri) [reply extensions/amp-list/0.1/test/test-amp-list.js]: Yep!

File: extensions/amp-list/0.1/test/test-amp-list.js (hunk 1 above)
Comment (samouri) [reply extensions/amp-list/0.1/test/test-amp-list.js]: Done

File: extensions/amp-list/0.1/amp-list.js (hunk 4 above)
Comment (samouri) [reply extensions/amp-list/0.1/amp-list.js]: Done

File: extensions/amp-list/0.1/amp-list.js (hunk 5 above)
Comment (samouri) [reply extensions/amp-list/0.1/amp-list.js]: Done

File: extensions/amp-list/0.1/test/test-amp-list.js (hunk 2 above)
Comment (samouri) [reply extensions/amp-list/0.1/test/test-amp-list.js]: I hope it is slightly more clear now. I await neither of them until the end now (where I await both)
